File: test/ut_tcp_client.cpp
#include <cstring>
#include <deque>

#include "common.hpp"
#include "Encoder.hpp"
#include "Packet.hpp"
#include "TcpClient.hpp"

#include "test_vectors.hpp"
#include "util.hpp"

int main(int argc, char ** argv) {
    if (3 > argc) {
        LOGE("Usage: %s <Server Address> <Server Port>\n", argv[0]);
        return 1;
    }

    std::unique_ptr<comm::TcpClient> pTcpClient = comm::TcpClient::create(
        argv[1], static_cast<uint16_t>(atoi(argv[2]))
    );

    if (!pTcpClient) {
        LOGE("Could not create Tcp Client which connects to %s/%s!\n", argv[1], argv[2]);
        return 1;
    }

    LOGI("Connected to (%s/%s)!\n", argv[1], argv[2]);

    for (size_t i = 0; i < vectors.size(); i++) {
        LOGI("[%lld (us)] Sent packet %zu (%zu bytes)\n", 
            static_cast<long long int>(get_elapsed_realtime_us()),
            i, vectors_sizes[i]
        );
#ifdef USE_RAW_POINTER
        pTcpClient->send(comm::Packet::create(vectors[i], vectors_sizes[i]));
#else
        std::unique_ptr<uint8_t[]> pdata(new uint8_t[vectors_sizes[i]]);
        memcpy(pdata.get(), vectors[i], vectors_sizes[i]);
        pTcpClient->send(comm::Packet::create(pdata, vectors_sizes[i]));
#endif  // USE_RAW_POINTER
    }

    LOGI("Sent %lu packets to server, press enter to check Rx Queue ...\n", vectors.size());
    getchar();

    std::deque<std::unique_ptr<comm::Packet>> pPackets;
    if (pTcpClient->recvAll(pPackets)) {
        if (test(pPackets)) {
            LOGI("-> Passed!\n");
        } else {
            LOGI("-> Failed!\n");
        }
    } else {
        LOGE("Rx Queue is empty!\n");
    }

    LOGI("Press enter to exit ...\n");
    getchar();

    return 0;
}

File: test/ut_tcp_client.py
import sys

sys.path.append('../wrapper')
import comm_wrapper
import ut_tcpip_ep

if __name__ == "__main__":
    if (3 > len(sys.argv)):
        print('Usage:', sys.argv[0], '<Server Address> <Server Port>\n')
        exit

    if (not comm_wrapper.load_library('config.ini')):
        print('Could not load shared library!')
        exit()

    print('Initializing TCP Client ...')
    if (comm_wrapper.comm_tcp_client_init(sys.argv[1], int(sys.argv[2]))):
        print('-> Successfully!')
    else:
        print('-> Failed!')

    if (ut_tcpip_ep.execute()):
        print('-> Passed!')
    else:
        print('-> Failed!')

    comm_wrapper.comm_deinit()

    print('Press enter to exit ...')
    input()
